src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1375 @@
+            // after this connection has been established. The chances are very
+            // small and the harm is very limited to the next fast connect
+            // attempt failing.
+            std::string ssid((char*)sdata->ssid, sdata->ssid_len);
+            last_sta_update(
+                ssid, staConnectPass_, sdata->authmode, sdata->channel);
+            wlan_connected(true, CONNECT_OK, ssid);
+            break;
+        }
+        case WIFI_EVENT_STA_DISCONNECTED:
+        {
+            // Note: It is possible to get here when we are already
+            //       "disconnected". Therefore, we should not assert on
+            //       connected_ == true. Every failed connection "attempt"
+            //       lands us here.
+            connected_ = false;
+            wifi_event_sta_disconnected_t *evdata =
+                (wifi_event_sta_disconnected_t *)data;
+            LOG(INFO, "wifi: STA disconnected ssid=%s reason %d rssi=%d",
+                evdata->ssid, evdata->reason, evdata->rssi);
+
+            bool try_fast_reconnect = false;
+            // For some reason the IP_EVENT_STA_LOST_IP event does not get
+            // delivered when there is an unexpected WiFi disconnect. I guess
+            // it is implied?
+            if (ipAcquiredSta_)
+            {
+                ipAcquiredSta_ = false; // Disconnect implies we lost IP.
+                try_fast_reconnect = true; // Try a fast reconnect.
+                mdns_disable_sta();
+                ip_acquired(IFACE_STA, false);
+            }
+            if (try_fast_reconnect || fastConnectOnlySta_)
+            {
+                // First reconnect attempt, or forced "fast" connect.
+                esp_wifi_connect();
+                LOG(INFO, "wifi: STA disconnected, fast reconnect attempt...");
+            }
+            else
+            {
+                // Successive reconnect attempts.
+                esp_wifi_scan_start(nullptr, false);
+                LOG(INFO, "wifi: STA disconnected, scanning...");
+            }
+            std::string ssid((char*)evdata->ssid, evdata->ssid_len);
+            wlan_connected(
+                false, connection_result_encode(evdata->reason), ssid);
+            break;
+        }
+        case WIFI_EVENT_SCAN_DONE:
+        {
+            /// @todo The newer IDF versions have a new api that can read
+            ///       AP records one at a time, which is much more memory
+            ///       friendly. Revisit this when updating the IDF version.
+            int idx = -1;
+            uint16_t number;
+            esp_wifi_scan_get_ap_num(&number);
+            LOG(INFO, "wifi: Scan done, number of records: %u.", number);
+            number = std::min(number, static_cast<uint16_t>(10));
+
+            // This takes a lot of memory. Not the best API design. See note
+            // above.
+            wifi_ap_record_t *ap_records = new wifi_ap_record_t[number];
```

Comment:
instead of dynamically allocatingthis memory I think we should consider allocating it statically. This avoids memory fragmentation. We're not super tight on RAM, so spending 1k of memory on this would be perfectly acceptable. We can use the constant I mentioned above. The callback thread ensures we will only ever use it one at a time.

Reply:
We will upgrade the IDF to resolve this.